### ISSUE TITLE
Assorted issues with context managers

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#9694]
**Describe the bug**

Context manager handling is inconsistent, and where issues with it are detected, the error messages do not correctly identify the root cause

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?code=JYWwDg9gTgLgBAQwM4E8B2BjYEBQOMA2yScA6tANbBoDmAogB4LgECmAXHnN4qpnABNWAMzgB9MQlZoYrKBIAUSVgWEBKOAFoAfHAByENB24A6Mzh690GQSPGTWDYDEXLVAGjgAqAYZcYEKGMIACMAK1YMGA0dfUNjODMTCx4hUQlpWXkxJRV1LV0DI3ZTc0s0%2B0dnVzzPHz8xAKCS0IiomML4ksTzPEJiOABBPgwASTQAN2AkYBCCFAAhKAgKaUZmMDZOS2RrW3SHGTka1Q644tLk8rsJKScXHLdhOt8ZRsDg8MjogvOEpO4AGI4KMhAgCPNPDAABbTOBIaEQACuBAEcBCrEQaDgcmWUDgMAgEE81AwEHACBgszYcAA7s5oYJKQg4CAIEICCluBUMkdsrlTr8iv8yqkbmIqg8BflYsLugC4MDQaxwZCCbCSAjkaj0ZiENjcdACUSSZhyWBKdTMfSYUyYCy2Ry8PgiEgSAA1abUxbLVZodYsDhc-b2KR8k4yzoXHomIEgsEQlBQjXwxEotEYrE4qB443EuCk82WubWhl2h3slTBnkOe4Rl4NJqfNo-WVdS5x5WqpPquFa9O6rOG-GE-OFilUkt0ssCZmsyuc64HTLHR55M5yy7V8WSiMb9sxvqukiDAhICDjT0zOY%2BlbSbZi5fhteCtvRpLOnDAA5oZisCRwAAvIBcAAEQSCACDUBIoEPtwwa0pQJAgeQUBULQAabKwChqMGUzXvMYghL60hAXAV7eksd7%2BkwgY4cGuyYGI1D4TeREkdiIHDNY4z4ZRHGYWw9EMSMIbMFBaA4UKXTBjsok2oyiFoUgcGWGpH5qTwjE2ApcB8WxxHUapmlxmAKBQMANDQvAQiyFEJDCgWJBoBA8CDLSUH2lOChkjm3zzAWQgyN%2BwCsGi0xIEirC4SZanAoitKsBMcjqpiw6sqwboIDQmJwoWflRJ4yAFvA-SgCQAAGtwedUYgVU5aAAOTwGAQTKDIskmRpJnadOtracxkzTAZHHGSZwJmRZVk2RAmVwC5NmsHZ8AwpScBVbWtX1dCxUwtayy0ASKBgKwnWaR%2BCFlkpFAqWdfWMvphGGX63QTeZlnWbYy0kDC0yeCESIrdCaU5kaICZUg2W5SQIDTGwCACNQNB3ZYwJlSAJ4jLxXo3lRfqCLNzmuQWgbgzI60ZHWdWeLSwPYs4BOZZ4YRIkg8D5UEUTzCjPDAgzvhzQt8KWkgwgoKlcBQEiwXg5LrAAI5IsAQRkzAJAQKIGJI4gcBkkcDDwJBv45fixUsm1KIwO4PNxueEvpXCvmc1b6KAxL4NZTljNIE17NBVSYsSwgURIuCTmRawVyxTwukDSxw1PaNtt-rwiAhBAyU23FBbYgjiNUoY6rIlZyZ9glP3A1YIDg1k4u6XtVj8LH6eZ9H43zcTWuHQESLTbGWdydY2AmFLknidQOFqEAA)

```python
import asyncio

class WorkingExample:

    async def __aenter__(self) -> None:  ...
    async def __aexit__(self, *dont_care: object) -> None:  ...
    def __enter__(self) -> None:  ...
    def __exit__(self, *dont_care: object) -> None:  ...


class AsyncInvisiblyBrokenExample:
    async def __aenter__(self) -> None:  ...
    def __aexit__(self, *dont_care: object) -> None:  ...  # Ideally, this should be an error too, incompatible with data model
    def __enter__(self) -> None:  ...
    def __exit__(self) -> None:  ...  # Ideally, this should be an error too, incompatible with data model


class VisiblyBrokenExample:
    def __aenter__(self) -> None:  ...  # Ideally, this should be an error too, incompatible with data model
    def __aexit__(self, *dont_care: object) -> None:  ...  # Ideally, this should be an error too, incompatible with data model
    def __enter__(self) -> None:  ...
    def __exit__(self) -> None:  ...

class AlsoInVisiblyBroken:
    def __enter__(self) -> None:  ...



if __name__ == "__main__":
    
    works = WorkingExample()
    visibly_broken = VisiblyBrokenExample()
    async_invisbly_broken = AsyncInvisiblyBrokenExample()

    async def amain() -> None:
        async with works:
            ...
        async with visibly_broken:
            # pyright detects None is not Awaitable (correctly identified issue)
            # however the error message is incorrect, as it claims `__await__` isn't present
            ...
        async with async_invisbly_broken:
            # pyright does not detect that `__aexit__` has the wrong type
            ...

    with works:
        with visibly_broken:  # pyright detects this, but the error message is misleading
            # claims AsyncInvisiblyBroken does not implement `__exit__`, when it does, just incorrectly
            # it does not satisfy the runtime requirements of being a context manager as a result,
            # so the error is correct, but the message doesn't identify the actual issue.
            with async_invisbly_broken:  # same as above
                # in addition though, this shows the asymmetery with the async with above
                # not being caught. 
                asyncio.run(amain())
```

This was discovered while typing some previously private code for public use with N:M concurrency for use with freethreading python builds.

edit: ignore the definition of `AlsoInVisiblyBroken`, it was there to demonstrate things that ended up redundant and couldd have been removed.

**Comment by erictraut** [comment]
An error should not be reported for the definition of `__aenter__` or `__aexit__` methods. These are not errors unless/until these methods are actually used. It would be inappropriate and inconsistent to report such an error when defined.

I think pyright's error message for the `visibly_broken` case is correct. The `__aenter__` method returns `None`, which is not awaitable. It's not awaitable because it doesn't conform to the `Awaitable` protocol. It doesn't conform to the `Awaitable` protocol because it doesn't have an `__await__` method. So that error message looks correct to me.

I've added an additional check to verify that the return result of `__aexit__` is awaitable.

I've also tweaked the error message for the case where the magic method is present but its signature is incompatible with the arguments being passed to it.

This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.392.
